git/revlist.go

```diff
@@ -117,3 +120,207 @@ func revListCallback(c *Client, opt RevListOptions, commits []CommitID, excludeL
+//  tag: tag1"
+var refNamesForCommit map[string]string
+
+// We have been asked to pre-compute the ref names for important
+//  commits as a way to improve performance.
+func buildRefNames(c *Client) error {
+	symRefsForRefName = make(map[string]string)
+	refNamesForCommit = make(map[string]string)
+
+	// TODO support more symbolic refs than just HEAD
+	headrefspec, err := SymbolicRefGet(c, SymbolicRefOptions{Short: true}, "HEAD")
+	if err != nil {
+		return err
+	}
+	symRefsForRefName[string(headrefspec)] = "HEAD"
+
+	branches, err := c.GetBranches()
+	if err != nil {
+		return err
+	}
+
+	for _, branch := range branches {
+		branchName := string(branch)
+
+		commitId, err := RefSpec(branchName).Value(c)
+		if err != nil {
+			//return err There are problems dealing with hierarchical branches
+			continue
+		}
+
+		branchName = branchName[len("refs/heads/"):]
+
+		if entry, ok := symRefsForRefName[branchName]; ok {
+			branchName = entry + " -> " + branchName
+		}
+
+		if entry, ok := refNamesForCommit[commitId]; ok {
+			refNamesForCommit[commitId] = entry + ", " + branchName
+		} else {
+			refNamesForCommit[commitId] = branchName
+		}
+	}
+
+	tagNames, err := TagList(c, TagOptions{}, []string{})
+	if err != nil {
+		return err
+	}
+
+	for _, tagName := range tagNames {
+		tag := RefSpec("refs/tags/" + tagName)
+		commitId, err := tag.Value(c)
+		if err != nil {
+			//return err There are problems dealing with hierarchical tags
+			continue
+		}
+
+		if entry, ok := refNamesForCommit[commitId]; ok {
+			refNamesForCommit[commitId] = entry + ", tag: " + tagName
+		} else {
+			refNamesForCommit[commitId] = "tag: " + tagName
+		}
+	}
+
+	return nil
+}
+
+// Since libgit is somewhat out of our control and we can't implement
+// a fmt.Stringer interface there, we use this helper.
+
+// Print commit using "medium" format
+func printCommitMedium(c *Client, cmt CommitID, format string) error {
+	author, err := cmt.GetAuthor(c)
+	if err != nil {
+		return err
+	}
+	fmt.Printf("commit %s\n", cmt)
+	if parents, err := cmt.Parents(c); len(parents) > 1 && err == nil {
+		fmt.Printf("Merge: ")
+		for i, p := range parents {
+			fmt.Printf("%s", p)
+			if i != len(parents)-1 {
+				fmt.Printf(" ")
+			}
+		}
+		fmt.Println()
+	}
+	date, err := cmt.GetDate(c)
+	if err != nil {
+		return err
+	}
+	fmt.Printf("Author: %v\nDate:   %v\n\n", author, date.Format("Mon Jan 2 15:04:05 2006 -0700"))
+	log.Printf("Commit %v\n", cmt)
+
+	msg, err := cmt.GetCommitMessage(c)
+	if err != nil {
+		return err
+	}
+	lines := strings.Split(strings.TrimSpace(msg.String()), "\n")
+	for _, l := range lines {
+		fmt.Printf("    %v\n", l)
+	}
+	fmt.Printf("\n")
+	return nil
+}
+
+func printCommitOneLine(c *Client, cmt CommitID, format string) error {
+	msg, err := cmt.GetCommitMessage(c)
+	if err != nil {
+		return err
+	}
+	// TODO figure out if merge parents should show up in this mode or not
+	title := strings.Split(strings.TrimSpace(msg.String()), "\n")[0]
```

Comment:
This needs a bounds check for empty messages, otherwise it'll panic.

Reply:
👍